libs/contrib/Data/Seq/Seq'.idr

```diff
@@ -0,0 +1,219 @@
+  then let (_, MkElem a) = lookupTree i t in Just a
+  else Nothing
+
+||| O(log(min(i, n-i))). Update the element at the specified position.
+||| If the position is out of range, the original sequence is returned.
+export
+adjust : (a -> a) -> Nat -> Seq' a -> Seq' a
+adjust f i s@(MkSeq' t) = if i < length' t
+  then MkSeq' $ adjustTree (const (map f)) i t
+  else s
+
+||| O(log(min(i, n-i))). Replace the element at the specified position.
+||| If the position is out of range, the original sequence is returned.
+export
+update : Nat -> a -> Seq' a -> Seq' a
+update i a t = adjust (const a) i t
+
+||| O(log(min(i, n-i))). Split a sequence at a given position.
+||| splitAt i s = (take i s, drop i s)
+export
+splitAt : Nat -> Seq' a -> (Seq' a, Seq' a)
+splitAt i s@(MkSeq' t) = if i < length' t
+  then let (l, r) = split i t 
+       in (MkSeq' l, MkSeq' r)
+  else (s, empty)
+
+||| O(log(min(i,n-i))). The first i elements of a sequence.
+||| If the sequence contains fewer than i elements, the whole sequence is returned.
+export 
+take : Nat -> Seq' a -> Seq' a
+take i seq = fst (splitAt i seq)
+
+||| O(log(min(i,n-i))). Elements of a sequence after the first i.
+||| If the sequence contains fewer than i elements, the empty sequence is returned.
+export 
+drop : Nat -> Seq' a -> Seq' a
+drop i seq = snd (splitAt i seq)
+
+||| Dump the internal structure of the finger tree.
+export
+show' : Show a => Seq' a -> String
+show' (MkSeq' tr) = showPrec Open tr
+
+public export
+implementation Eq a => Eq (Seq' a) where
+  MkSeq' x == MkSeq' y = x == y
+
+public export
+implementation Ord a => Ord (Seq' a) where
+  compare (MkSeq' x) (MkSeq' y) = compare x y
+
+public export
+implementation Functor Seq' where
+  map f (MkSeq' tr) = MkSeq' (map (map f) tr)
+
+public export
+implementation Foldable Seq' where
+  foldr f z (MkSeq' tr) = foldr (f . unElem) z tr
+
+  foldl f z (MkSeq' tr) = foldl (\acc, (MkElem elem) => f acc elem) z tr
+
+  toList (MkSeq' tr) = toList' tr
+
+  null (MkSeq' Empty) = True
+  null _ = False
+
+public export
+implementation Traversable Seq' where
+  traverse f (MkSeq' tr) = MkSeq' <$> traverse (map MkElem . f . unElem) tr
+
+public export
+implementation Show a => Show (Seq' a) where
+  showPrec p = showPrec p . toList
+
+public export
+implementation Zippable Seq' where
+  zipWith f (MkSeq' x) (MkSeq' y) = MkSeq' (zipWith' f x y)
+
+  zipWith3 f (MkSeq' x) (MkSeq' y) (MkSeq' z) = MkSeq' (zipWith3' f x y z) 
+
+  unzipWith f (MkSeq' zs) = let (xs, ys) = unzipWith' f zs in (MkSeq' xs, MkSeq' ys)
+      
+  unzipWith3 f (MkSeq' ws) = let (xs, ys, zs) = unzipWith3' f ws in (MkSeq' xs, MkSeq' ys, MkSeq' zs)
+
+public export
+implementation Semigroup (Seq' a) where
+  (<+>) = (++)
+
+public export
+implementation Monoid (Seq' a) where
+  neutral = empty
+
+||| This implementation is differnt from that of Seq.
+public export
+implementation Applicative Seq' where
+  pure = singleton
+  fs <*> xs = concat $ map (\f => map f xs) fs
```

Comment:
`concat . map` = `foldMap` (unless it's not `Prelude.concat`)

Reply:
Ok, I forgot.